packages/webpack-cli/lib/groups/basicResolver.js

```diff
@@ -12,7 +17,41 @@ const WEBPACK_OPTION_FLAGS = core
         return result;
     }, []);
 
-function resolveArgs(args) {
+const PRODUCTION = 'production';
+const DEVELOPMENT = 'development';
+
+/*
+    Mode priority:
+        - Mode flag
+        - Mode from config
+        - Mode form NODE_ENV
+    */
+/**
+ *
+ * @param {string} mode - mode flag value
+ * @param {Object} configObject - contains relevant loaded config
+ */
+const assignMode = (mode, configObject) => {
+    const {
+        env: { NODE_ENV },
+    } = process;
+    const { mode: configMode } = configObject;
+    let finalMode;
+    if (mode) {
+        finalMode = mode;
+    } else if (configMode) {
+        finalMode = configMode;
+    } else if (NODE_ENV && (NODE_ENV === PRODUCTION || NODE_ENV === DEVELOPMENT)) {
+        finalMode = NODE_ENV;
+    } else {
+        finalMode = PRODUCTION;
+    }
```

Comment:
Ideally we should remove it, because we set production by default, but it is out of scope webpack-cli, otherwise webpack will not throw warning on empty mode value

Reply:
yep lets remove

Reply:
Also let's add test for this to avoid regression in future

Reply:
done

Reply:
I think we should do it separately, a lot of tests are breaking